In [1]:
pip install --upgrade pip

  Using cached pip-24.2-py3-none-any.whl.metadata (3.6 kB)
Using cached pip-24.2-py3-none-any.whl (1.8 MB)
  Attempting uninstall: pip
    Found existing installation: pip 24.1.2
    Uninstalling pip-24.1.2:
      Successfully uninstalled pip-24.1.2


In [2]:
pip install tensorflow

In [3]:
import tensorflow as tf
print("version: ", tf.__version__)

version:  2.17.0


Load the Dataset:

We are using the [MNIST](https://www.kaggle.com/datasets/hojjatk/mnist-dataset) dataset in Keras that has 70,000 images, with 60,000 for training and 10,000 for testing.


In [10]:
mnist = tf.keras.datasets.mnist

In [11]:
(x_train, y_train),(x_test, y_test) = mnist.load_data()

In [12]:
x_train, x_test = x_train/255.0,x_test/255.0

Build the machine learning model:

In [13]:
model = tf.keras.models.Sequential([
    tf.keras.layers.Flatten(input_shape=(28,28)),
    tf.keras.layers.Dense(128, activation='relu'),
    tf.keras.layers.Dropout(0.2),
    tf.keras.layers.Dense(10)
])

/usr/local/lib/python3.10/dist-packages/keras/src/layers/reshaping/flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


In [14]:
predictions = model(x_train[:1]).numpy()
predictions

array([[ 0.19422668, -0.10509518, -0.43796447, -0.25392753, -0.3691678 ,
         0.19872832,  0.4098698 ,  0.6935816 ,  0.06132449,  0.5058366 ]],
      dtype=float32)

In [15]:
tf.nn.softmax(predictions).numpy()

array([[0.10400602, 0.07710182, 0.05527151, 0.06643969, 0.05920786,
        0.10447527, 0.12903605, 0.17136635, 0.09106254, 0.14203289]],
      dtype=float32)

Define the loss function:

In [16]:
loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)

In [17]:
loss_fn(y_train[:1], predictions).numpy()

2.2588048

In [18]:
model.compile(optimizer='adam',
              loss=loss_fn,
              metrics=['accuracy'])

Train and evaluate the model:

In [19]:
model.fit(x_train, y_train, epochs=5)

Epoch 1/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step - accuracy: 0.8601 - loss: 0.4850
Epoch 2/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 8s 3ms/step - accuracy: 0.9555 - loss: 0.1523
Epoch 3/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step - accuracy: 0.9665 - loss: 0.1091
Epoch 4/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.9727 - loss: 0.0895
Epoch 5/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step - accuracy: 0.9776 - loss: 0.0726


In [20]:
model.evaluate(x_test,y_test,verbose=2)

313/313 - 1s - 2ms/step - accuracy: 0.9760 - loss: 0.0719


[0.07185832411050797, 0.9760000109672546]

The image classifier now is trained to have ~98% accuracy